pkg/service/model_builder.go

```diff
@@ -80,17 +91,73 @@ type defaultModelBuilder struct {
 	enableBackendSG          bool
 	disableRestrictedSGRules bool
 
+	stackGlobalCache map[core.StackID]core.Stack
+
 	clusterName         string
 	vpcID               string
 	defaultTags         map[string]string
 	externalManagedTags sets.String
 	defaultSSLPolicy    string
 	defaultTargetType   elbv2model.TargetType
 	enableIPTargetType  bool
+
+	client client.Client
+
+	initialized bool
+	lock        sync.RWMutex
 }
 
 func (b *defaultModelBuilder) Build(ctx context.Context, service *corev1.Service) (core.Stack, *elbv2model.LoadBalancer, bool, error) {
-	stack := core.NewDefaultStack(core.StackID(k8s.NamespacedName(service)))
+	// Initialize the global cache if not initialized
+	if !b.initialized {
+		// if not initialized, we need to build the global cache based on existing services
+		var serviceList corev1.ServiceList
+		if err := b.client.List(ctx, &serviceList); err != nil {
+			return nil, nil, false, err
+		}
+		for _, svc := range serviceList.Items {
+			if svc.Annotations[LoadBalancerStackKey] != "" && svc.DeletionTimestamp.IsZero() {
+				stackID := core.StackID(types.NamespacedName{
+					Namespace: "stack",
+					Name:      svc.Annotations[LoadBalancerStackKey],
+				})
+				if b.stackGlobalCache[stackID] == nil {
+					b.lock.Lock()
+					b.stackGlobalCache[stackID] = core.NewDefaultStack(stackID)
+					b.lock.Unlock()
+				}
+				b.lock.Lock()
+				b.stackGlobalCache[stackID].AddService(&svc)
+				b.lock.Unlock()
```

Comment:
The read should also be protected here.
```suggestion
				b.lock.Lock()
				if b.stackGlobalCache[stackID] == nil {
					b.stackGlobalCache[stackID] = core.NewDefaultStack(stackID)
				}
				b.stackGlobalCache[stackID].AddService(&svc)
				b.lock.Unlock()
```

Reply:
Actually, is the `lock` here even necessary? You have the other lock and that should take care of this, too, right?

Reply:
Yeah I will change to lock the whole snippet. It is necessary since another lock onlys locks per stack, not globally.